dot/sync/worker_test.go

```diff
@@ -9,53 +9,84 @@ import (
 	"time"
 
 	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/lib/common/variadic"
 	"github.com/libp2p/go-libp2p/core/peer"
 	"github.com/stretchr/testify/require"
-	"go.uber.org/mock/gomock"
 )
 
 func TestWorker(t *testing.T) {
 	peerA := peer.ID("peerA")
-	ctrl := gomock.NewController(t)
-
-	reqMaker := NewMockRequestMaker(ctrl)
-	reqMaker.EXPECT().
-		Do(peerA, nil, gomock.AssignableToTypeOf((*network.BlockResponseMessage)(nil))).
-		DoAndReturn(func(_, _, _ any) any {
-			time.Sleep(2 * time.Second)
+	m := uint32(60)
+	blockReq := &network.BlockRequestMessage{
+		RequestedData: 1,
+		StartingBlock: variadic.Uint32OrHash{},
+		Direction:     3,
+		Max:           &m,
+	}
+	requestCounter := 0
+	reqMaker := fakeReqMaker{
+		doFunc: func(id peer.ID, req network.Message, resp network.ResponseMessage) error {
+			// assert that what is added as a task in the channel
+			// queue, this will be used in the requests
+			require.Equal(t, peerA, id)
+			require.Equal(t, blockReq, req)
+			resp = new(network.BlockResponseMessage)
+			requestCounter++
 			return nil
-		}).
-		Times(2).
-		Return(nil)
+		},
+	}
 
 	sharedGuard := make(chan struct{}, 1)
 	w := newWorker(peerA, sharedGuard, reqMaker)
 
 	wg := sync.WaitGroup{}
-	queue := make(chan *syncTask, 2)
+	queue := make(chan *syncTask, 1) // 1 means that maximum 1 worker can send request at a given time
 
-	wg.Add(1)
+	// run two workers, but they shouldn't work concurrently,
+	// because sharedGuard is buffered channel with capacity
+	wg.Add(2)
+	go w.run(queue, &wg)
 	go w.run(queue, &wg)
 
 	resultCh := make(chan *syncTaskResult)
 	defer close(resultCh)
-
 	queue <- &syncTask{
+		request:  blockReq,
 		resultCh: resultCh,
 	}
-
 	queue <- &syncTask{
+		request:  blockReq,
 		resultCh: resultCh,
 	}
 
+	// we are waiting 500 ms to guarantee that workers had time to read sync tasks from the queue
+	// and send the request. With this assertion we can be sure that even that we start 2 workers
+	// only one of them is working and sent a requests
 	time.Sleep(500 * time.Millisecond)
-	require.Equal(t, 1, len(sharedGuard))
-	<-resultCh
+	require.Equal(t, 1, requestCounter)
```

Comment:
In the current test code we also ensure that only one worker is "working" by checking the `sharedGuard` channel length: `require.Equal(t, 1, len(sharedGuard))`. That's way simpler and is relying on a channel used internally by the worker, which acquires the `lock` and releases it to the next. So I would suggest to leave this check as it is.

Reply:
Sorry that I didn't explain why I changed that. I will do this now. The old test actually doesn't guarantee that only one worker is working at a given time. For example, if someone changes the production code to:
```go
func executeRequest(who peer.ID, requestMaker network.RequestMaker,
	task *syncTask, sharedGuard chan struct{}) {
	//defer func() {
	//	<-sharedGuard
	//}()

	if len(sharedGuard) < 1 {
		sharedGuard <- struct{}{}
	}

	//sharedGuard <- struct{}{}
        ...
}
```

The test will pass because len(sharedGuard) is 1. 

The new test case will fail with:
```text
Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestWorker
```
Also, in the old test case, we ran only one worker which is not enough to guarantee that more than one worker can be running at the moment. I think it is better for at least 2 workers to be run at the same time and assert in the tests that only one of them can work at a given time.
 
Even if my proposal for the test is not good, I think that we should find a way to test the number of workers better. For example, we can run two workers, check the length of the channel when workers are active/work, and,  check the channel again after all workers finish with their work that its length is 0. In this way, we can be more confident that only one worker is running at a given time. What do you think about that?:

```go
  ...
  go w.run(queue, &wg)
  go w.run(queue, &wg)<

  // send two tasks in the queue
  queue <- syncTask{...}
  queue <- syncTask{...}
  
 require.Equal(t, 1, len(sharedGuard)) // assert that only one of the workers is running  at the moment
 
 ...
 
 wg.Wait() 
 require.Equal(t, 0, len(sharedGuard)) // assert that the buffered channel is empty because all the workers finished their work.
  
```

Reply:
if you make such changes the code will be incorrect resulting in failures.

For example, if I have 100 concurrent process spawned almost at the same time and those process will call the function `executeRequest`, the very first that reaches the line `sharedGuard <- struct {}` will "acquire" the lock, so the other 99 process which reaches that line will be blocked, that happens because our `sharedGuard` channel is a buffered channel of size 1, if we increase the size to 4, then the first 4 process which reaches that line will "acquire" and the other 96 process will wait... So after one process finishes its execution, I mean, exec the request, got the response and decode the response it will return, this will trigger the `defer` function within the `executeRequest` function, and that defer fun releases the acquired lock `<-sharedGuard`, as soon as the process, which acquired the lock, releases it then 1 from the 99 remaining process will be able to acquire it.

So, given my prev explanation, we don't need to check the sharedGuard len to acquire a lock, we can directly exec `sharedGuard <- struct{}` and if the lock is free then it will successfully acquire it and proceed, but if it is already locked then it will not allow to proceed, blocking until it is free.
```
sharedGuard := make(chan struct{}, 1)
...
func executeRequest(...) {
    ...
    defer func() { <-sharedGuard }
    sharedGuard <- struct {}
    ...
}
```

That's how we use channels to serve as a more specialized mutex, given that by using mutex I can guarantee that only one actor access the resource, and using this sharedGuard I can define how many actors can access a resource/execute an action.

So, there is no need to change the production code (in the actual context, even though we are evaluating a refactor in this package)

You can add in this PR more workers, like I've suggested in this comment https://github.com/ChainSafe/gossamer/pull/4033#discussion_r1638189639, and the tests will pass.

Reply:
Hey @EmilGeorgiev! Thanks for the clarification, I just notice that the purpose of the PR is to ensure that it covers the case when shared guard is not impl properly. Good catch actually, so I would suggest an approach to that:

In the current test setup our shared guard has a slot for a single execution, so we should fail if two workers try to call the `reqMaker.Do` mock method, so we can use a helper channel to do that:
```
// acquireOrFail is a test channel used to
// ensure the shared guard is working properly
// should have the same len as the shared guard
acquireOrFail := make(chan struct{}, 1)

// in the mocked method we use the select in order to acquire the helper channel
// if we could not acquire (meaning that another process acquire it and is using it) we fall in 
// the default case where we fail the test
reqMaker.EXPECT().
		Do(peerA, nil, gomock.AssignableToTypeOf((*network.BlockResponseMessage)(nil))).
		DoAndReturn(func(_, _, _ any) any {
			select {
			case acquireOrFail <- struct{}{}:
				defer func() {
					<-acquireOrFail // release once it finishes
				}()
			default:
				t.Errorf("should acquire the channel, othewise the shared guard is not working")
			}
...
use the same `acquireOrFail` channel for the other mocked `Do` call
```

That should be ok, to ensure that only the amount of allowed slots are executing the requests concurrently and no other call is made.

Again, thanks for noticing that!

Reply:
I like the idea of the helper channel. 

And the library "go.uber.org/mock/gomock" is very useful. In the beginning, when I used the fake implementation of RequestMaker I didn't realize that the library provided the ability to write such custom logic.

I pushed the new changes